File: backend/model_library/network/retry.py
"""Retry logic with exponential backoff for network operations.

Provides configurable retry behavior with exponential backoff and jitter
to handle transient network failures gracefully.
"""

from __future__ import annotations

import asyncio
import random
import time
from dataclasses import dataclass, field
from typing import Any, Callable, TypeVar

from backend.logging_config import get_logger

logger = get_logger(__name__)

T = TypeVar("T")


@dataclass
class RetryConfig:
    """Configuration for retry behavior.

    Attributes:
        max_attempts: Maximum number of attempts (including first try)
        base_delay: Initial delay in seconds
        max_delay: Maximum delay cap in seconds
        exponential_base: Base for exponential calculation
        jitter: Whether to add randomness to delays
    """

    max_attempts: int = 3
    base_delay: float = 1.0
    max_delay: float = 60.0
    exponential_base: float = 2.0
    jitter: bool = True


@dataclass
class RetryStats:
    """Statistics from a retry operation.

    Attributes:
        attempts: Number of attempts made
        total_delay: Total time spent waiting between retries
        success: Whether the operation ultimately succeeded
        last_exception: The last exception raised, if any
    """

    attempts: int = 0
    total_delay: float = 0.0
    success: bool = False
    last_exception: BaseException | None = None

    def record_attempt(self, delay: float = 0.0) -> None:
        """Record an attempt with optional delay."""
        self.attempts += 1
        self.total_delay += delay

    def record_success(self) -> None:
        """Mark operation as successful."""
        self.success = True

    def record_failure(self, exc: BaseException) -> None:
        """Record a failure with exception."""
        self.success = False
        self.last_exception = exc


class RetryError(Exception):
    """Raised when all retry attempts are exhausted.

    Attributes:
        stats: Statistics from the retry operation
    """

    def __init__(self, message: str, stats: RetryStats) -> None:
        super().__init__(message)
        self.stats = stats

    def __str__(self) -> str:
        base_msg = super().__str__()
        if self.stats.last_exception:
            return f"{base_msg}: {self.stats.last_exception}"
        return base_msg


def calculate_backoff(attempt: int, config: RetryConfig) -> float:
    """Calculate backoff delay for a given attempt.

    Uses exponential backoff with optional jitter.

    Args:
        attempt: The current attempt number (0-indexed)
        config: Retry configuration

    Returns:
        Delay in seconds before next retry
    """
    # Exponential backoff: base_delay * exponential_base ^ attempt
    delay = config.base_delay * (config.exponential_base**attempt)

    # Cap at max_delay
    delay = min(delay, config.max_delay)

    # Add jitter if enabled (random value between 0 and delay)
    if config.jitter and delay > 0:
        delay = random.uniform(0, delay * 2)  # noqa: S311
        delay = min(delay, config.max_delay)

    return delay


def should_retry(
    exc: BaseException,
    retryable_exceptions: list[type[BaseException]] | None = None,
    predicate: Callable[[BaseException], bool] | None = None,
) -> bool:
    """Determine if an exception should trigger a retry.

    Args:
        exc: The exception that was raised
        retryable_exceptions: List of exception types that are retryable
        predicate: Custom function to determine if exception is retryable

    Returns:
        True if the operation should be retried
    """
    # Predicate takes precedence
    if predicate is not None:
        return predicate(exc)

    # Check against exception types
    if retryable_exceptions:
        return isinstance(exc, tuple(retryable_exceptions))

    return False


def retry(
    fn: Callable[..., T],
    config: RetryConfig,
    *args: Any,
    retryable_exceptions: list[type[BaseException]] | None = None,
    predicate: Callable[[BaseException], bool] | None = None,
    on_retry: Callable[[int, BaseException, float], None] | None = None,
    **kwargs: Any,
) -> tuple[T, RetryStats]:
    """Execute a function with retry logic.

    Args:
        fn: Function to execute
        config: Retry configuration
        *args: Positional arguments for fn
        retryable_exceptions: Exception types that trigger retry
        predicate: Custom predicate to determine if retry should occur
        on_retry: Callback called before each retry (attempt, exception, delay)
        **kwargs: Keyword arguments for fn

    Returns:
        Tuple of (result, stats)

    Raises:
        RetryError: If all attempts are exhausted
        Exception: Non-retryable exceptions are raised immediately
    """
    stats = RetryStats()

    for attempt in range(config.max_attempts):
        try:
            stats.record_attempt()
            result = fn(*args, **kwargs)
            stats.record_success()
            return result, stats

        except BaseException as exc:  # noqa: generic-exception
            stats.record_failure(exc)

            # Check if we should retry
            if not should_retry(exc, retryable_exceptions, predicate):
                raise

            # Check if we have attempts remaining
            if attempt + 1 >= config.max_attempts:
                logger.warning(
                    "Retry exhausted after %d attempts: %s",
                    stats.attempts,
                    exc,
                )
                raise RetryError(
                    f"All {config.max_attempts} attempts failed",
                    stats,
                ) from exc

            # Calculate and apply backoff
            delay = calculate_backoff(attempt, config)
            stats.total_delay += delay

            logger.debug(
                "Attempt %d failed, retrying in %.2fs: %s",
                attempt + 1,
                delay,
                exc,
            )

            if on_retry:
                on_retry(attempt + 1, exc, delay)

            time.sleep(delay)

    # Should not reach here, but satisfy type checker
    raise RetryError(f"All {config.max_attempts} attempts failed", stats)


async def retry_async(
    fn: Callable[..., Any],
    config: RetryConfig,
    *args: Any,
    retryable_exceptions: list[type[BaseException]] | None = None,
    predicate: Callable[[BaseException], bool] | None = None,
    on_retry: Callable[[int, BaseException, float], None] | None = None,
    **kwargs: Any,
) -> tuple[Any, RetryStats]:
    """Execute an async function with retry logic.

    Args:
        fn: Async function to execute
        config: Retry configuration
        *args: Positional arguments for fn
        retryable_exceptions: Exception types that trigger retry
        predicate: Custom predicate to determine if retry should occur
        on_retry: Callback called before each retry (attempt, exception, delay)
        **kwargs: Keyword arguments for fn

    Returns:
        Tuple of (result, stats)

    Raises:
        RetryError: If all attempts are exhausted
        Exception: Non-retryable exceptions are raised immediately
    """
    stats = RetryStats()

    for attempt in range(config.max_attempts):
        try:
            stats.record_attempt()
            result = await fn(*args, **kwargs)
            stats.record_success()
            return result, stats

        except BaseException as exc:  # noqa: generic-exception
            stats.record_failure(exc)

            # Check if we should retry
            if not should_retry(exc, retryable_exceptions, predicate):
                raise

            # Check if we have attempts remaining
            if attempt + 1 >= config.max_attempts:
                logger.warning(
                    "Async retry exhausted after %d attempts: %s",
                    stats.attempts,
                    exc,
                )
                raise RetryError(
                    f"All {config.max_attempts} attempts failed",
                    stats,
                ) from exc

            # Calculate and apply backoff
            delay = calculate_backoff(attempt, config)
            stats.total_delay += delay

            logger.debug(
                "Async attempt %d failed, retrying in %.2fs: %s",
                attempt + 1,
                delay,
                exc,
            )

            if on_retry:
                on_retry(attempt + 1, exc, delay)

            await asyncio.sleep(delay)

    # Should not reach here, but satisfy type checker
    raise RetryError(f"All {config.max_attempts} attempts failed", stats)

File: rust/crates/pumas-core/src/network/retry.rs
//! Retry logic with exponential backoff and jitter.
//!
//! Provides configurable retry behavior for network operations with:
//! - Exponential backoff (delay doubles each attempt)
//! - Optional jitter to prevent thundering herd
//! - Customizable retry predicates
//! - Statistics tracking

use rand::Rng;
use std::future::Future;
use std::time::Duration;
use tracing::{debug, warn};

/// Configuration for retry behavior.
#[derive(Debug, Clone)]
pub struct RetryConfig {
    /// Maximum number of attempts (including the first one).
    pub max_attempts: u32,
    /// Initial delay between retries.
    pub base_delay: Duration,
    /// Maximum delay cap.
    pub max_delay: Duration,
    /// Exponential base (typically 2.0 for doubling).
    pub exponential_base: f64,
    /// Whether to add random jitter to delays.
    pub jitter: bool,
}

impl Default for RetryConfig {
    fn default() -> Self {
        Self {
            max_attempts: 3,
            base_delay: Duration::from_secs(1),
            max_delay: Duration::from_secs(60),
            exponential_base: 2.0,
            jitter: true,
        }
    }
}

impl RetryConfig {
    /// Create a new retry config with default values.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set the maximum number of attempts.
    pub fn with_max_attempts(mut self, attempts: u32) -> Self {
        self.max_attempts = attempts;
        self
    }

    /// Set the base delay.
    pub fn with_base_delay(mut self, delay: Duration) -> Self {
        self.base_delay = delay;
        self
    }

    /// Set the maximum delay cap.
    pub fn with_max_delay(mut self, delay: Duration) -> Self {
        self.max_delay = delay;
        self
    }

    /// Enable or disable jitter.
    pub fn with_jitter(mut self, jitter: bool) -> Self {
        self.jitter = jitter;
        self
    }

    /// Calculate the delay for a given attempt number (0-indexed).
    pub fn calculate_delay(&self, attempt: u32) -> Duration {
        // Exponential backoff: base * (exponential_base ^ attempt)
        let multiplier = self.exponential_base.powi(attempt as i32);
        let delay_secs = self.base_delay.as_secs_f64() * multiplier;
        let capped_secs = delay_secs.min(self.max_delay.as_secs_f64());

        let final_secs = if self.jitter {
            // Decorrelated jitter: multiply delay by random factor between 0.5 and 1.5
            // This keeps the average delay the same while adding randomness to prevent
            // thundering herd, without allowing near-zero delays
            let mut rng = rand::rng();
            let jitter_factor = rng.random_range(0.5..1.5);
            (capped_secs * jitter_factor).min(self.max_delay.as_secs_f64())
        } else {
            capped_secs
        };

        Duration::from_secs_f64(final_secs)
    }
}

/// Statistics about a retry operation.
#[derive(Debug, Clone, Default)]
pub struct RetryStats {
    /// Number of attempts made.
    pub attempts: u32,
    /// Total delay accumulated.
    pub total_delay: Duration,
    /// Whether the operation ultimately succeeded.
    pub success: bool,
    /// Last error message if failed.
    pub last_error: Option<String>,
}

/// Retry an async operation with exponential backoff.
///
/// # Arguments
///
/// * `config` - Retry configuration
/// * `operation` - Async function that returns a Result
/// * `should_retry` - Predicate to determine if an error is retryable
///
/// # Returns
///
/// A tuple of (Result, RetryStats)
pub async fn retry_async<F, Fut, T, E>(
    config: &RetryConfig,
    mut operation: F,
    should_retry: impl Fn(&E) -> bool,
) -> (Result<T, E>, RetryStats)
where
    F: FnMut() -> Fut,
    Fut: Future<Output = Result<T, E>>,
    E: std::fmt::Display,
{
    let mut stats = RetryStats::default();

    for attempt in 0..config.max_attempts {
        stats.attempts = attempt + 1;

        match operation().await {
            Ok(value) => {
                stats.success = true;
                if attempt > 0 {
                    debug!("Operation succeeded after {} attempts", attempt + 1);
                }
                return (Ok(value), stats);
            }
            Err(e) => {
                stats.last_error = Some(e.to_string());

                // Check if we should retry
                if !should_retry(&e) {
                    debug!("Error is not retryable: {}", e);
                    return (Err(e), stats);
                }

                // Check if we have more attempts
                if attempt + 1 >= config.max_attempts {
                    warn!(
                        "All {} retry attempts exhausted. Last error: {}",
                        config.max_attempts, e
                    );
                    return (Err(e), stats);
                }

                // Calculate and apply delay
                let delay = config.calculate_delay(attempt);
                stats.total_delay += delay;

                warn!(
                    "Attempt {}/{} failed: {}. Retrying in {:?}",
                    attempt + 1,
                    config.max_attempts,
                    e,
                    delay
                );

                tokio::time::sleep(delay).await;
            }
        }
    }

    // Should not reach here, but just in case
    unreachable!("Retry loop should have returned")
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicU32, Ordering};
    use std::sync::Arc;

    #[test]
    fn test_delay_calculation_no_jitter() {
        let config = RetryConfig::new()
            .with_base_delay(Duration::from_secs(1))
            .with_jitter(false);

        // First attempt: 1 * 2^0 = 1s
        assert_eq!(config.calculate_delay(0), Duration::from_secs(1));
        // Second attempt: 1 * 2^1 = 2s
        assert_eq!(config.calculate_delay(1), Duration::from_secs(2));
        // Third attempt: 1 * 2^2 = 4s
        assert_eq!(config.calculate_delay(2), Duration::from_secs(4));
    }

    #[test]
    fn test_delay_capped_at_max() {
        let config = RetryConfig::new()
            .with_base_delay(Duration::from_secs(10))
            .with_max_delay(Duration::from_secs(30))
            .with_jitter(false);

        // 10 * 2^3 = 80s, but capped at 30s
        assert_eq!(config.calculate_delay(3), Duration::from_secs(30));
    }

    #[test]
    fn test_delay_with_jitter() {
        let config = RetryConfig::new()
            .with_base_delay(Duration::from_secs(2))
            .with_jitter(true);

        // With decorrelated jitter, delay should be between 0.5x and 1.5x the base
        // For attempt 0 with base 2s: expected range is 1s to 3s
        for _ in 0..20 {
            let delay = config.calculate_delay(0);
            // Jitter factor is 0.5 to 1.5, so delay should be 1s to 3s
            assert!(
                delay >= Duration::from_secs(1) && delay <= Duration::from_secs(3),
                "Delay {:?} should be between 1s and 3s",
                delay
            );
        }

        // For attempt 1 with base 2s: 2 * 2^1 = 4s, range is 2s to 6s
        for _ in 0..20 {
            let delay = config.calculate_delay(1);
            assert!(
                delay >= Duration::from_secs(2) && delay <= Duration::from_secs(6),
                "Delay {:?} should be between 2s and 6s",
                delay
            );
        }
    }

    #[tokio::test]
    async fn test_retry_succeeds_first_try() {
        let config = RetryConfig::new().with_max_attempts(3);

        let (result, stats) =
            retry_async(&config, || async { Ok::<_, String>(42) }, |_: &String| true).await;

        assert_eq!(result.unwrap(), 42);
        assert_eq!(stats.attempts, 1);
        assert!(stats.success);
    }

    #[tokio::test]
    async fn test_retry_succeeds_after_failures() {
        let config = RetryConfig::new()
            .with_max_attempts(3)
            .with_base_delay(Duration::from_millis(10))
            .with_jitter(false);

        let counter = Arc::new(AtomicU32::new(0));
        let counter_clone = counter.clone();

        let (result, stats) = retry_async(
            &config,
            || {
                let counter = counter_clone.clone();
                async move {
                    let count = counter.fetch_add(1, Ordering::SeqCst);
                    if count < 2 {
                        Err("temporary failure".to_string())
                    } else {
                        Ok(42)
                    }
                }
            },
            |_: &String| true,
        )
        .await;

        assert_eq!(result.unwrap(), 42);
        assert_eq!(stats.attempts, 3);
        assert!(stats.success);
    }

    #[tokio::test]
    async fn test_retry_exhausted() {
        let config = RetryConfig::new()
            .with_max_attempts(3)
            .with_base_delay(Duration::from_millis(10))
            .with_jitter(false);

        let (result, stats) = retry_async(
            &config,
            || async { Err::<i32, _>("always fails".to_string()) },
            |_: &String| true,
        )
        .await;

        assert!(result.is_err());
        assert_eq!(stats.attempts, 3);
        assert!(!stats.success);
        assert_eq!(stats.last_error, Some("always fails".to_string()));
    }

    #[tokio::test]
    async fn test_retry_non_retryable_error() {
        let config = RetryConfig::new().with_max_attempts(3);

        let (result, stats) = retry_async(
            &config,
            || async { Err::<i32, _>("permanent failure".to_string()) },
            |e: &String| !e.contains("permanent"),
        )
        .await;

        assert!(result.is_err());
        assert_eq!(stats.attempts, 1); // Only one attempt for non-retryable
        assert!(!stats.success);
    }
}
